src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+            middleCurrent.prev = nodeByIndex.prev;
+            nodeByIndex.prev = middleCurrent;
+            middleCurrent.next = nodeByIndex;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t:list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index == 0) {
+            final T oldValue = head.item;
+            Node<T> newHead = new Node<>(null, value, head.next);
+            newHead.next = head.next;
+            head.next.prev = newHead;
+            head = newHead;
+            return oldValue;
+        }
+        Node<T> nodeToChange = getNode(index);
+        Node<T> node = new Node<>(nodeToChange.prev, value, nodeToChange.next);
+        nodeToChange.prev.next = node;
+        nodeToChange.next.prev = node;
+        return nodeToChange.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T result = node.item;
+        unlink(node);
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            Node<T> node = getNode(i);
+            if (node.item == object || node.item != null && node.item.equals(object)) {
+                unlink(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void unlink(Node<T> node) {
+        if (node == head) {
+            head = node.next;
+            if (head != null) {
+                head.prev = null;
+            } else {
+                tail = null;
+            }
+        } else if (node == tail) {
+            tail = node.prev;
+            if (tail != null) {
+                tail.next = null;
+            } else {
+                head = null;
+            }
+        } else {
+            node.next.prev = node.prev;
+            node.prev.next = node.next;
+        }
+        size--;
+    }
+
+    private static class Node<E> {
+        private final E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        private Node(Node<E> prev, E element, Node<E> next) {
+            this.prev = prev;
+            this.item = element;
+            this.next = next;
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
```

Comment:
Please add logic to iterate from tail when index is closer to the end of list
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa